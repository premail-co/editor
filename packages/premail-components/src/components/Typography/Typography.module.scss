@mixin plainText {
  text-align: left;
  font-family: "Open Sans";
  font-style: normal;
  font-weight: 400;
  font-size: 16px;
  line-height: 20px;
  letter-spacing: 0px;
  color: var(--typography-color);
  opacity: 1;
}

.h1 {
  text-align: left;
  font-family: "PT Sans";
  font-style: normal;
  font-weight: 700;
  font-size: 36px;
  line-height: 52px;
  letter-spacing: 0px;
  color: var(--typography-color);
  opacity: 1;
}

.h2 {
  text-align: left;
  font-family: "PT Sans";
  font-style: normal;
  font-weight: 700;
  font-size: 32px;
  line-height: 41px;
  letter-spacing: 0px;
  color: var(--typography-color);
  opacity: 1;
}

.h3 {
  text-align: left;
  font-family: "PT Sans";
  font-style: normal;
  font-weight: 700;
  font-size: 28px;
  line-height: 36px;
  letter-spacing: 0px;
  color: var(--typography-color);
  opacity: 1;
}

.h4 {
  text-align: left;
  font-family: "PT Sans";
  font-style: normal;
  font-weight: 700;
  font-size: 24px;
  line-height: 31px;
  letter-spacing: 0px;
  color: var(--typography-color);
  opacity: 1;
}

.h5 {
  text-align: left;
  font-family: "Open Sans";
  font-style: normal;
  font-weight: 700;
  font-size: 20px;
  line-height: 26px;
  letter-spacing: 0px;
  color: var(--typography-color);
  opacity: 1;
}

.h6 {
  text-align: left;
  font-family: "Open Sans";
  font-style: normal;
  font-weight: 700;
  font-size: 16px;
  line-height: 22px;
  letter-spacing: 0px;
  color: var(--typography-color);
  opacity: 1;
}

.p {
  @include plainText;
}

.li {
  @include plainText;

  // &::before{
  // content: "\2022";  /* Add content: \2022 is the CSS Code/unicode for a bullet */
  // color: var(--typography-color);; /* Change the color */
  // font-weight: bold; /* If you want it to be bold */
  // display: inline-block; /* Needed to add space between the bullet and the text */
  // width: 16px; /* Also needed for space (tweak if needed) */
  // margin-left: -16px; /* Also needed for space (tweak if needed) */
  // }
}

.strike {
  @include plainText;
  text-decoration: line-through;
}

@mixin link {
  text-align: left;
  font-family: "Open Sans";
  font-style: normal;
  font-weight: 400;
  font-size: 16px;
  line-height: 20px;
  letter-spacing: 0px;
  color: var(--premail-link);
  opacity: 1;
}

.link {
  @include link;
}
