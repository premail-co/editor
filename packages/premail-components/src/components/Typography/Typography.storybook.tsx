import React from "react";
import reactElementToJSXString from "react-element-to-jsx-string";
import { Typography } from ".";
import { Surface } from "../Surface/Surface";
// This default export determines where you story goes in the story list
export default {
  title: "Typography",
  component: Typography,
};

const CodeBlock = (
  props: React.PropsWithChildren<{ style?: React.CSSProperties }>
) => {
  return (
    <div
      style={{
        padding: "40px",
        whiteSpace: "nowrap",
        backgroundColor: "var(--premail-gray-0)",
      }}
    >
      <pre
        style={{
          whiteSpace: "pre-wrap",
          wordBreak: "break-all",
          backgroundColor: "var(--premail-gray-0)",
          margin: "none",
        }}
      >
        <code style={{ color: "var(--premail-black)", whiteSpace: "pre-wrap" }}>
          {props.children}
        </code>
      </pre>
    </div>
  );
};

const InlineCode = (props: React.PropsWithChildren<{}>) => {
  return (
    <code
      style={{
        color: "var(--premail-black)",
        whiteSpace: "pre-wrap",
        backgroundColor: "var(--premail-gray-0)",
      }}
    >
      {props.children}
    </code>
  );
};

export const Overview = () => {
  return (
    <div style={{ maxWidth: 900, margin: "auto", padding: "0px 5px" }}>
      <Typography variant={"h1"} renderElement="h1">
        Typography
      </Typography>

      <Typography variant="p1" renderElement="p">
        Typograhy is a set of styles and appearance of headers, paragraphs and
        general text content. The Typography component can render several
        typography variants using the <InlineCode>variant</InlineCode> prop and
        can define the final rendered element by the{" "}
        <InlineCode>renderElement</InlineCode> prop.
      </Typography>

      <Typography variant={"h2"} renderElement="h2">
        Usage
      </Typography>
      <Typography variant="p1" renderElement="p">
        To begin using the <InlineCode>Typography</InlineCode> component, import
        the following:
      </Typography>
      <CodeBlock>{`import { Typography } from "@premail/components";`}</CodeBlock>
      <Typography variant={"h2"} renderElement="h2">
        Variants
      </Typography>
      <Typography variant="p1" renderElement="p">
        The following variants are defined in{" "}
        <InlineCode>@premail/components</InlineCode> Typography:{" "}
        <InlineCode>h1</InlineCode>, <InlineCode>h2</InlineCode>,{" "}
        <InlineCode>h3</InlineCode>, <InlineCode>h4</InlineCode>,{" "}
        <InlineCode>h5</InlineCode>, <InlineCode>h6</InlineCode>,{" "}
        <InlineCode>p1</InlineCode>, <InlineCode>p2</InlineCode> and{" "}
        <InlineCode>p3</InlineCode>
      </Typography>
      <Surface
        border
        style={{
          padding: "40px",
          display: "flex",
          flexDirection: "column",
        }}
      >
        <Typography variant={"h1"} disableMargins>
          Header H1
        </Typography>
        <Typography variant={"h2"} disableMargins>
          Header H2
        </Typography>
        <Typography variant={"h3"} disableMargins>
          Header H3
        </Typography>
        <Typography variant={"h4"} disableMargins>
          Header H4
        </Typography>
        <Typography variant={"h5"} disableMargins>
          Header H5
        </Typography>
        <Typography variant={"h6"} disableMargins>
          Header H6
        </Typography>
        <Typography variant={"p1"} disableMargins>
          Text p1
        </Typography>
        <Typography variant={"p2"} disableMargins>
          Text p2
        </Typography>
        <Typography variant={"p3"} disableMargins>
          Text p3
        </Typography>
      </Surface>
      <CodeBlock>
        {`<Surface
  border
  style={{
    padding: "40px",
    display: "flex",
    flexDirection: "column",
  }}
>
  <Typography variant={"h1"} disableMargins> Header H1 </Typography>
  <Typography variant={"h2"} disableMargins> Header H2 </Typography>
  <Typography variant={"h3"} disableMargins> Header H3 </Typography>
  <Typography variant={"h4"} disableMargins> Header H4 </Typography>
  <Typography variant={"h5"} disableMargins> Header H5 </Typography>
  <Typography variant={"h6"} disableMargins> Header H6 </Typography>
  <Typography variant={"p1"} disableMargins> Text p1 </Typography>
  <Typography variant={"p2"} disableMargins> Text p2 </Typography>
  <Typography variant={"p3"} disableMargins> Text p3 </Typography>
</Surface>`}
      </CodeBlock>

      <Typography variant={"h2"} renderElement="h2">
        Decoration props
      </Typography>
      <Typography variant="p1" renderElement="p">
        For each variant, typography can be decorated using the{" "}
        <InlineCode>bold</InlineCode>, <InlineCode>italic</InlineCode>,{" "}
        <InlineCode>link</InlineCode> and <InlineCode>strikethrough</InlineCode>{" "}
        props.
      </Typography>

      <Surface
        border
        style={{
          padding: "40px",
          display: "flex",
          flexDirection: "column",
        }}
      >
        <Typography variant={"p1"} bold>
          The quick brown fox jumps over the lazy dog.
        </Typography>
        <Typography variant={"p1"} italic>
          The quick brown fox jumps over the lazy dog.
        </Typography>
        <Typography variant={"p1"} link>
          The quick brown fox jumps over the lazy dog.
        </Typography>
        <Typography variant={"p1"} strikethrough>
          The quick brown fox jumps over the lazy dog.
        </Typography>
      </Surface>
      <CodeBlock>
        {`<Surface
  border
  style={{
    padding: "40px",
    display: "flex",
    flexDirection: "column",
  }}
>
  <Typography variant={"p1"} bold>
    The quick brown fox jumps over the lazy dog.
  </Typography>
  <Typography variant={"p1"} italic>
    The quick brown fox jumps over the lazy dog.
  </Typography>
  <Typography variant={"p1"} link>
    The quick brown fox jumps over the lazy dog.
  </Typography>
  <Typography variant={"p1"} strikethrough>
    The quick brown fox jumps over the lazy dog.
  </Typography>
</Surface>`}
      </CodeBlock>
      <Typography variant={"h2"} renderElement="h2">
        Render Elements
      </Typography>
      <Typography variant="p1" renderElement="p">
        The <InlineCode>Typography</InlineCode> component can also render its
        set of styles in a custom element determined by the{" "}
        <InlineCode>renderElement</InlineCode> prop. This allows to decouple the
        design set of typography styles from the final rendered tag which can be
        useful to keep your html content{" "}
        <a href="https://developer.mozilla.org/en-US/docs/Glossary/Semantics#semantics_in_html">
          semantic
        </a>
        .
      </Typography>
      <Surface
        border
        style={{
          padding: "40px",
          display: "flex",
          flexDirection: "column",
        }}
      >
        <Typography variant={"p1"} renderElement="p" disableMargins>
          I render in a p tag.
        </Typography>
        <Typography variant={"p1"} renderElement="span" disableMargins>
          I render in a span tag.
        </Typography>

        <Typography variant={"p1"} renderElement="h1" disableMargins>
          I render in an h1 tag.
        </Typography>

        <Typography variant={"p1"} renderElement="h2" disableMargins>
          I render in an h2 tag.
        </Typography>

        <Typography variant={"p1"} renderElement="h3" disableMargins>
          I render in an h3 tag.
        </Typography>

        <Typography variant={"p1"} renderElement="h5" disableMargins>
          I render in an h5 tag.
        </Typography>

        <Typography variant={"p1"} renderElement="h5" disableMargins>
          I render in an h5 tag.
        </Typography>

        <Typography variant={"p1"} renderElement="h6" disableMargins>
          I render in an h6 tag.
        </Typography>
      </Surface>
      <CodeBlock>
        {`<Surface
  border
  style={{
    padding: "40px",
    display: "flex",
    flexDirection: "column",
  }}
>
  <Typography variant={"p1"} renderElement="p" disableMargins>
    I render in a p tag.
  </Typography>
  <Typography variant={"p1"} renderElement="span" disableMargins>
    I render in a span tag.
  </Typography>

  <Typography variant={"p1"} renderElement="h1" disableMargins>
    I render in an h1 tag.
  </Typography>

  <Typography variant={"p1"} renderElement="h2" disableMargins>
    I render in an h2 tag.
  </Typography>

  <Typography variant={"p1"} renderElement="h3" disableMargins>
    I render in an h3 tag.
  </Typography>

  <Typography variant={"p1"} renderElement="h5" disableMargins>
    I render in an h5 tag.
  </Typography>

  <Typography variant={"p1"} renderElement="h5" disableMargins>
    I render in an h5 tag.
  </Typography>

  <Typography variant={"p1"} renderElement="h6" disableMargins>
    I render in an h6 tag.
  </Typography>
</Surface>`}
      </CodeBlock>
    </div>
  );
};
