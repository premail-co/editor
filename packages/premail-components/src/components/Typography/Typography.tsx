// React
import React from "react";
import { TypographyVariant_ENUM } from "./constants";
// type import
import type { IOverridable, IIdentifiable, IStylable } from "../../types";
// Styles
import styles from "./Typography.module.scss";

interface ITypographyProps extends IIdentifiable, IStylable, IOverridable {
  variant: TypographyVariant_ENUM;
}

function Typography(props: React.PropsWithChildren<ITypographyProps>) {
  const className =
    props.override != null
      ? `${styles[props.variant]} ${props.override}`
      : styles[props.variant];
  return (
    <span className={className} style={props.style}>
      {props.children}
    </span>
  );
}
export default Typography;
export { TypographyVariant_ENUM };
export type { ITypographyProps };
