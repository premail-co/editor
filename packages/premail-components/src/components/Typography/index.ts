//Typography
export { Typography } from "./Typography";
export type { ITypographyProps } from "./Typography";
