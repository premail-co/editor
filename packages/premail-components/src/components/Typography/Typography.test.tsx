import React from "react";
import ReactDOM from "react-dom/client";
import { act } from "react-dom/test-utils";
import Typography, { TypographyVariant_ENUM } from "./Typography";

describe("Typography tests", () => {
  let container: HTMLElement | null = document.createElement("div");
  let root: ReactDOM.Root | null = null;

  beforeEach(() => {
    const newContainer = document.createElement("div");
    document.body.appendChild(newContainer);
    act(() => {
      root = ReactDOM.createRoot(newContainer);
    });
    container = newContainer;
  });

  afterEach(() => {
    if (container == null || root == null) return;
    act(() => {
      if (root == null) return;
      root.unmount();
    });
    document.body.removeChild(container);
    container = null;
  });

  it("Should Match H1 Snapshots", () => {
    act(() => {
      root?.render(
        <Typography variant={TypographyVariant_ENUM.h1}>
          Typography H1 Variant
        </Typography>
      );
    });
    let tree = container?.innerHTML;
    expect(tree).toMatchSnapshot();
  });
  it("should match H2 snapshot", () => {
    act(() => {
      root?.render(
        <Typography variant={TypographyVariant_ENUM.h2}>
          Typography H2 Variant
        </Typography>
      );
    });
    let tree = container?.innerHTML;
    expect(tree).toMatchSnapshot();
  });

  it("Should Match H3 snapshots", () => {
    act(() => {
      root?.render(
        <Typography variant={TypographyVariant_ENUM.h3}>
          Typography H3 Variant
        </Typography>
      );
    });
    let tree = container?.innerHTML;
    expect(tree).toMatchSnapshot();
  });

  it("Should Match H4 snapshots", () => {
    act(() => {
      root?.render(
        <Typography variant={TypographyVariant_ENUM.h4}>
          Typography H4 Variant
        </Typography>
      );
    });
    let tree = container?.innerHTML;
    expect(tree).toMatchSnapshot();
  });

  it("Should Match H5 snapshots", () => {
    act(() => {
      root?.render(
        <Typography variant={TypographyVariant_ENUM.h5}>
          Typography H5 Variant
        </Typography>
      );
    });
    let tree = container?.innerHTML;
    expect(tree).toMatchSnapshot();
  });

  it("Should Match H6 snapshots", () => {
    act(() => {
      root?.render(
        <Typography variant={TypographyVariant_ENUM.h6}>
          Typography H6 Variant
        </Typography>
      );
    });
    let tree = container?.innerHTML;
    expect(tree).toMatchSnapshot();
  });

  it("Should Match p snapshots", () => {
    act(() => {
      root?.render(
        <Typography variant={TypographyVariant_ENUM.p}>
          Typography p Variant
        </Typography>
      );
    });
    let tree = container?.innerHTML;
    expect(tree).toMatchSnapshot();
  });

  it("Should Match Link snapshots", () => {
    act(() => {
      root?.render(
        <Typography variant={TypographyVariant_ENUM.link}>
          Typography link Variant
        </Typography>
      );
    });
    let tree = container?.innerHTML;
    expect(tree).toMatchSnapshot();
  });

  it("Should render overriding classNames", () => {
    act(() => {
      root?.render(
        <Typography
          variant={TypographyVariant_ENUM.link}
          override={"testClass"}
        >
          Typography link Variant
        </Typography>
      );
    });

    const el = container?.querySelector(
      `[class="${TypographyVariant_ENUM.link} testClass"]`
    );

    expect(el).not.toBeNull();
  });

  it("Should render style objects", () => {
    const style = { color: "red" };

    act(() => {
      root?.render(
        <Typography variant={TypographyVariant_ENUM.link} style={style}>
          Typography link Variant
        </Typography>
      );
    });
    const el = container?.querySelector<HTMLElement>(
      `[class="${TypographyVariant_ENUM.link}"]`
    );

    if (el == null) fail();

    expect(el.style.color).toBe("red");
  });
});
