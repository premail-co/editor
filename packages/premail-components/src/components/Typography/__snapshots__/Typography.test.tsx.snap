// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography tests Should Match H1 Snapshots 1`] = `"<span class=\\"root h1\\">Typography H1 Variant</span>"`;

exports[`Typography tests Should Match H3 snapshots 1`] = `"<span class=\\"root h3\\">Typography H3 Variant</span>"`;

exports[`Typography tests Should Match H4 snapshots 1`] = `"<span class=\\"root h4\\">Typography H4 Variant</span>"`;

exports[`Typography tests Should Match H5 snapshots 1`] = `"<span class=\\"root h5\\">Typography H5 Variant</span>"`;

exports[`Typography tests Should Match H6 snapshots 1`] = `"<span class=\\"root h6\\">Typography H6 Variant</span>"`;

exports[`Typography tests Should Match p snapshots 1`] = `"<span class=\\"root p1\\">Typography p Variant</span>"`;

exports[`Typography tests should match H2 snapshot 1`] = `"<span class=\\"root h2\\">Typography H2 Variant</span>"`;
